picard/acoustid/manager.py

```diff
@@ -39,62 +39,92 @@ def __init__(self, fingerprint, duration, orig_recordingid=None, recordingid=Non
+    MAX_ATTEMPTS = 5
+    # In case of Payload Too Large error, batch size is reduced by this factor
+    # and a retry occurs
+    BATCH_SIZE_REDUCTION_FACTOR = 0.7
 
     def __init__(self, acoustid_api):
         super().__init__()
-        self._fingerprints = {}
+        self._submissions = {}
         self._acoustid_api = acoustid_api
 
     def add(self, file, recordingid):
         if not file.acoustid_fingerprint or not file.acoustid_length:
             return
         puid = file.metadata['musicip_puid']
-        self._fingerprints[file] = Submission(file.acoustid_fingerprint, file.acoustid_length, recordingid, recordingid, puid)
+        self._submissions[file] = Submission(file.acoustid_fingerprint, file.acoustid_length, recordingid, recordingid, puid)
         self._check_unsubmitted()
 
     def update(self, file, recordingid):
-        submission = self._fingerprints.get(file)
+        submission = self._submissions.get(file)
         if submission is None:
             return
         submission.recordingid = recordingid
         self._check_unsubmitted()
 
     def remove(self, file):
-        if file in self._fingerprints:
-            del self._fingerprints[file]
+        if file in self._submissions:
+            del self._submissions[file]
         self._check_unsubmitted()
 
     def is_submitted(self, file):
-        submission = self._fingerprints.get(file)
+        submission = self._submissions.get(file)
         if submission:
-            return not submission.recordingid or submission.orig_recordingid == submission.recordingid
+            return submission.is_submitted()
         return True
 
     def _unsubmitted(self):
-        for file, submission in self._fingerprints.items():
-            if submission.recordingid and submission.orig_recordingid != submission.recordingid:
+        for file, submission in self._submissions.items():
+            if not submission.is_submitted():
                 yield (file, submission)
 
     def _check_unsubmitted(self):
         enabled = next(self._unsubmitted(), None) is not None
         self.tagger.window.enable_submit(enabled)
 
     def submit(self):
+        self.max_batch_size = self.MAX_PAYLOAD
         submissions = list(self._unsubmitted())
         if not submissions:
             self._check_unsubmitted()
             return
         log.debug("AcoustID: submitting total of %d fingerprints...", len(submissions))
         self._batch_submit(submissions)
 
+    def _batch(self, submissions):
+        batch = []
+        remaining = []
+        batch_size = 0
+        for file, submission in submissions:
+            if submission.attempts < self.MAX_ATTEMPTS:
+                batch_size += len(submission)
+                if batch_size < self.max_batch_size:
+                    submission.attempts += 1
+                    batch.append((file, submission))
+                    continue
+            remaining.append((file, submission))
```

Comment:
It is probably not necessary to iterate over all submissions every time. After batch size hits max it is only appending the remaining, so we don't really need to check the submission attempts and length again. I would instead of the continue do a break and then add all the not yet processed items in submissions to the remaining list (this probably needs keeping count of the already processed ones).

What I don't quite get is how we should handle the submissions which reached their submission attempt limit. As I understand it they stay in the manager forever, getting iterated over and over again. Maybe we should reset the submission attempts after one submission cycle finished?

Reply:
> What I don't quite get is how we should handle the submissions which reached their submission attempt limit. As I understand it they stay in the manager forever, getting iterated over and over again. Maybe we should reset the submission attempts after one submission cycle finished?

Well,if a set of submissions reaches max attempts, those weren't submitted, keeping it in the manager let the user attempts a (manual) retry, in case cause of failure was temporary.
Really not sure what else we can do.

Reply:
> It is probably not necessary to iterate over all submissions every time. After batch size hits max it is only appending the remaining, so we don't really need to check the submission attempts and length again. I would instead of the continue do a break and then add all the not yet processed items in submissions to the remaining list (this probably needs keeping count of the already processed ones).

What do you think about 8cff4372559a51af70e7682cf478b5dddacec310 to address this?

Reply:
Looks good to me.

But I still don't get the submission.attempts handling :)

Later manual retry is fine for me. Causes for reached max. attempts might be interent connectivity or server issues, and retrying later can help.

But as I see it the attempt count for a submission never gets reset. So even on a later retry to submit the fingerprints the submission count would be ready at .sc attempts. Or do I miss something?

Reply:
> But as I see it the attempt count for a submission never gets reset.

Ah, you're right.
What about resetting them in https://github.com/metabrainz/picard/blob/8cff4372559a51af70e7682cf478b5dddacec310/picard/acoustid/manager.py#L105-L112